### PR TITLE
CSV Parser - improved

### PR DESCRIPTION
**Pull request by DSDubit**
Utility which allows the caller to specify a format of expected CSV data, then parse a file which matches that format in a type-safe way, including converting the input to common (int, bool, float, enum) value types, and custom conversions (including e.g. value range constraints), then validating the entire document.

Now includes the option of outputting the contents in CSV format, and writing new records to the document at runtime (i.e. you can write validated CSV files too now).

Deluxe(TM)

### CODE REVIEW

**Comment by DSDubit** [comment]
Use case from TOPPS:

- define some custom types which extend the basic CSV.Type, but with input pre-manipulation, value range limits, etc
- create a dictionary of field names to expected types
- parse a CSV file with this format
- optionally reject empty values (e.g. cardID), or support them as default(T) (e.g. player_name)

        private static CSV.Type cardValue = new CSV.Type<float>((str) =>
		{
			return float.Parse(str.Replace(" Million", string.Empty));
		}, allowEmptyValues: true);

		private static CSV.Type cardStat = new CSV.Types.Int(allowEmptyValues: true, minValue: 0, maxValue: 101);

		private static CSV.Type cardFoil = new CSV.Type<CardDefines.FoilType>((str) =>
		{
			return (CardDefines.FoilType)System.Enum.Parse(typeof(CardDefines.FoilType), "Foil_" + str);
		}, allowEmptyValues: true);

                ingestionModel = new Dictionary<string, CSV.Type>
		{
			// Generic
			{ CSVKey.card_id.ToString(), new CSV.Types.String() },
			{ CSVKey.type.ToString(), new CSV.Types.Enum<CardDefines.CardType>() },

			// Visual
			{ CSVKey.club.ToString(), new CSV.Types.String(allowEmptyValues: true) },

			// Player
			{ CSVKey.position.ToString(), new CSV.Types.Enum<CardDefines.Role>(allowEmptyValues: true) },
			{ CSVKey.player_name.ToString(), new CSV.Types.String(allowEmptyValues: true) },
			{ CSVKey.country_flag.ToString(), new CSV.Types.String(allowEmptyValues: true) },
			{ CSVKey.value.ToString(), cardValue },
			{ CSVKey.defence.ToString(), cardStat },
			{ CSVKey.attack.ToString(), cardStat },
			{ CSVKey.stars.ToString(), new CSV.Types.Int(allowEmptyValues: true, minValue: 0, maxValue: 5) },
			{ CSVKey.speed.ToString(), cardStat },
			{ CSVKey.tackle.ToString(), cardStat },
			{ CSVKey.power.ToString(), cardStat },
			{ CSVKey.shoot.ToString(), cardStat },
			{ CSVKey.skill.ToString(), cardStat },
			{ CSVKey.pass.ToString(), cardStat },
			{ CSVKey.foil.ToString(), cardFoil }
		};

                var doc = CSV.Parse(csvFile.text, new CSV.Format(ingestionModel), allowUnexpectedFields: false);

**Comment by DSDubit** [comment]
Write example from TOPPS:

                    var csvDocument = new CSV.WriteableDocument(csvFormat);
			foreach (var cardDefinition in response.cards)
			{
				csvDocument.AddRecord(ServerDefinitionToCSVValues(cardDefinition, csvFormat));
			}
			File.WriteAllText(path, csvDocument.Output());

**Comment by kkjamie** [comment]
Looks like good work, but so far only 1 project has had need for this. I'm wondering if duck is the correct place for this, or perhaps it's own repo

File: Utils/CSV.cs [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+
+			internal string GetFieldName(int index)
+			{
+				if (orderedKeys != null)
+				{
+					return (orderedKeys[index]);
+				}
+				else
+				{
+					return (format.Keys.ToArray())[index];
```
Comment (HarryHodgson) [inline Utils/CSV.cs]: `format.ElemantAt(index)`

Comment (kkjamie) [reply Utils/CSV.cs]: I'd recommend againt pulling things out of dictionaries by index. It's not guaranteed to be the same order you added them (although in my experience it always is, I think it's just an implementation detail). Also why is orderedKeys possible to be null ?

Comment (kkjamie) [reply Utils/CSV.cs]: Just for reference from: https://msdn.microsoft.com/en-us/library/xfhwa508.aspx 

> For purposes of enumeration, each item in the dictionary is treated as a KeyValuePair<TKey, TValue> structure representing a value and its key. The order in which the items are returned is undefined.

**Comment by FenrirWillow** [comment]
My two cents for does this need to live in DUCK or not - I don't understand the argument for it not being there. I personally view DUCK as our toolkit for doing various things easier. The most common against I have heard up until now has been "But I don't use X so why does this library pull it in". Don't necessarily see why that is an issue. Dead code optimization will remove it from the binary anyway. The engine frameworks for Unreal and Unity already have a huge surface area, compared to that, DUCK's API surface is tiny. Not sure why that would be a factor as well.

Those are my thoughts, obviously the Unity team gets to use this more often than I do, but I think a little more openness on what gets into DUCK can lead to some improvements in the amount of tools available for the teams to play around with.

**Comment by danhalldubit** [comment]
@FenrirWillow I agree with you, but something worth keeping in mind is that the more we add the more upkeep there is to do every Unity version update. It can get quite hard to maintain and start causing problems for projects, especially when the original developers or use cases are gone (e.g. when @Rollo-Liu had to just strip out the old duck audio system entirely)

**Pull request merged**